beacon-chain/sync/pending_blobs_queue.go

```diff
@@ -0,0 +1,144 @@
+package sync
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/prysmaticlabs/prysm/v4/beacon-chain/core/feed"
+	statefeed "github.com/prysmaticlabs/prysm/v4/beacon-chain/core/feed/state"
+	"github.com/prysmaticlabs/prysm/v4/config/params"
+	"github.com/prysmaticlabs/prysm/v4/encoding/bytesutil"
+	eth "github.com/prysmaticlabs/prysm/v4/proto/prysm/v1alpha1"
+	"github.com/prysmaticlabs/prysm/v4/time/slots"
+)
+
+// processPendingBlobs listens for state changes and handles pending blobs.
+func (s *Service) processPendingBlobs() {
+	s.waitForChainStart()
+
+	eventFeed := make(chan *feed.Event, 1)
+	sub := s.cfg.stateNotifier.StateFeed().Subscribe(eventFeed)
+	defer sub.Unsubscribe()
+
+	// Initialize the cleanup ticker
+	cleanupTicker := slots.NewSlotTicker(s.cfg.chain.GenesisTime(), params.BeaconConfig().SecondsPerSlot/2)
+
+	for {
+		select {
+		case <-s.ctx.Done():
+			return
+		case <-sub.Err():
+			return
+		case e := <-eventFeed:
+			s.handleEvent(s.ctx, e)
+		case <-cleanupTicker.C():
+			if s.pendingBlobSidecars == nil {
+				return
+			}
+			s.pendingBlobSidecars.cleanup()
+		}
+	}
+}
+
+// handleEvent processes incoming events.
+func (s *Service) handleEvent(ctx context.Context, e *feed.Event) {
+	if e.Type == statefeed.BlockProcessed {
+		s.handleNewBlockEvent(ctx, e)
+	}
+}
+
+// handleNewBlockEvent processes blobs when a parent block is processed.
+func (s *Service) handleNewBlockEvent(ctx context.Context, e *feed.Event) {
+	data, ok := e.Data.(*statefeed.BlockProcessedData)
+	if !ok {
+		return
+	}
+	s.processBlobsFromSidecars(ctx, data.SignedBlock.Block().ParentRoot())
+}
+
+// processBlobsFromSidecars processes blobs for a given parent root.
+func (s *Service) processBlobsFromSidecars(ctx context.Context, parentRoot [32]byte) {
+	blobs := s.pendingBlobSidecars.pop(parentRoot)
+	for _, blob := range blobs {
+		if err := s.validateAndReceiveBlob(ctx, blob); err != nil {
+			log.WithError(err).Error("Failed to validate blob in pending queue")
+		}
+	}
+}
+
+// validateAndReceiveBlob validates and processes a blob.
+func (s *Service) validateAndReceiveBlob(ctx context.Context, blob *eth.SignedBlobSidecar) error {
```

Comment:
do we need to name it validate and Receive? and not just receive with a comment saying it's validating.